nibabel/funcs.py

```diff
@@ -88,43 +88,67 @@ def squeeze_image(img):
                  img.extra)
 
 
-def concat_images(images, check_affines=True):
-    ''' Concatenate images in list to single image, along last dimension
+def concat_images(images, check_affines=True, axis=None):
+    ''' Concatenate images in list to single image, along specified dimension
 
     Parameters
     ----------
     images : sequence
-       sequence of ``SpatialImage`` or of filenames\s
+       sequence of ``SpatialImage`` or filenames\s
     check_affines : {True, False}, optional
        If True, then check that all the affines for `images` are nearly
        the same, raising a ``ValueError`` otherwise.  Default is True
-
+    axis : None or int, optional
+        If None, concatenates on the 4th dimension.
+        If not None, concatenates on the specified dimension [-2 to 3).
     Returns
     -------
     concat_img : ``SpatialImage``
        New image resulting from concatenating `images` across last
        dimension
     '''
+
     n_imgs = len(images)
-    img0 = images[0]
-    is_filename = False
-    if not hasattr(img0, 'get_data'):
-        img0 = load(img0)
-        is_filename = True
-    i0shape = img0.shape
-    affine = img0.affine
-    header = img0.header
-    out_shape = (n_imgs, ) + i0shape
-    out_data = np.empty(out_shape)
+    if n_imgs == 0:
+        raise ValueError('Cannot concatenate an empty list of images.')
+
     for i, img in enumerate(images):
-        if is_filename:
+        if not hasattr(img, 'get_data'):
             img = load(img)
-        if check_affines:
-            if not np.all(img.affine == affine):
-                raise ValueError('Affines do not match')
-        out_data[i] = img.get_data()
-    out_data = np.rollaxis(out_data, 0, len(i0shape)+1)
-    klass = img0.__class__
+
+        if i == 0:  # first image, initialize data from loaded image
+            affine = img.affine
+            header = img.header
+            klass = img.__class__
+
+            if axis is None:  # collect images in output array for efficiency
+                out_shape = (n_imgs, ) + img.shape[:3]
+                out_data = np.empty(out_shape)
+            else:  # collect images in list for use with np.concatenate
+                out_data = [None] * n_imgs
+
+        elif check_affines and not np.all(img.affine == affine):
+            raise ValueError('Affines do not match')
+
+        # Special case for 4D image with size[3] == 1; reshape to work!
```

Comment:
So - I still think this should not be special cased - because it's difficult to explain and the caller can just pass axis=-1 if they want this behavior.